### ISSUE TITLE
Secure keystroke forwarding for password manager autotype

### ISSUE THREAD

**Issue by fullTalgoRythm** [QubesOS/qubes-issues#5080]
**The problem you're addressing (if any)**
When accessing some web sites there can be up to 3 pages of randomised and partial secret pass-phrase / memorable data required to login. This can normally be handled by auto type in the password manager. This will however obviously not work where the password manager is running in a separate domain as there is currently no way to forward the keystrokes or select the correct autotype phrase. There are also some applications that either disable or break clipboard functionality which would benefit from auotype support.

I would also like to understand if you are open to improving the usability for ordinary end user requirements before trying to submit any patches myself.

**Describe the solution you'd like**
At the moment the first field of the logon page/prompt is selected in the relevant application. A keyboard shortcut is then selected (the pw manager has an associated shortcut). The pw manager then determines the active window and displays a choice of auto type selections available based on the active window name. This can handle multiple logon screens, partial pw input (often required from more than 1 field i.e. not just the main pw) including drop down selection of individual pw characters.

There are potentially multiple ways to achieve this but it would be nice if this could be as simple and secure as possible.

An example

There is a domain such as dom-pw that runs the password manager. When a user definable keyboard shortcut is selected it is intercepted in dom0 (similar to the clipboard function). This would then trigger the password manager window (presumably you would need a settings that indicates the domain can handle the specific keyboard shortcut). The actual exchange of data could probably be facilitated by RPC (I could offer to extend keepassxc for compatibility) or you could start & activate a dummy application in dom-pw  with the correct window name which has the sole purpose of capturing the required keyboard strokes. dom0 would know the correct window to send the keystrokes to (this could be on a timeout if you wanted to be paranoid). I don't believe this is worse than providing clipboard or file copy support as the results of key insertion are normally quite visible and the data rate is normally quite low. The pw manager must also surely be considered a trusted piece of software as it already has access to your pws.

An alternative might be that I extend keepassxc to display the selection of all available autotype for an entry and copy the autotype data to the clipboard in a yet to be agreed format. If you were willing to then provide a clipboard -> keyboard function (via keyboard shortcut) this could then send the clipboard data as keystrokes (including tab, up / down arrows etc). Such a function would be useful anywhere the clipboard refused to work (which is often during password entry). This would be less straight forward for the end user but perhaps it is an easier feature request?

**Where is the value to a user, and who might that user be?**
I believe this type of functionality would help to make the OS more enterprise friendly. I believe there is a wide range of sysadmin / it support users that have to deal with non clipboard friendly applications not to mention most typical end users probably have to deal with some logon pages like I described earlier. I believe this would improve the UX and encourage the use of strong passwords. The ability to automate a reliable mechanism to create, update and enter strong passwords on sites that require convoluted security (as they consider themselves high risk) would surely be a benefit given the overall theme of the project.

**Describe alternatives you've considered**
Perhaps it is possible to run the pw manager in dom0 but that would not be best practice (although if you don't trust your pw manager you obviously shouldn't use it at all). The most common choices like keepassx and keepassxc only have basic auto type so they would need to be extended in which case it would be a good opportunity to provide extensions compatible with Qubes. This would also highlight the OS to a key demographic that is probably not aware of its existence (and hence may not currently be asking why Qubes doesn't have this feature).

**Comment by andrewdavidwong** [comment]
> I believe this type of functionality would help to make the OS more enterprise friendly. I believe there is a wide range of sysadmin / it support users that have to deal with non clipboard friendly applications not to mention most typical end users probably have to deal with some logon pages like I described earlier. I believe this would improve the UX and encourage the use of strong passwords. The ability to automate a reliable mechanism to create, update and enter strong passwords on sites that require convoluted security (as they consider themselves high risk) would surely be a benefit given the overall theme of the project.

Yes, I can see the value in this. If you look at password managers like LastPass, for example, a lot of their value proposition has recently moved into automating password management tasks _other_ than generation and storage. I can imagine this sort of functionality being appealing to enterprise clients and less technical end users. (Our value proposition would, of course, be that the passwords and the manager itself exist in a securely isolated domain rather than in the web browser.)

**Comment by marmarek** [comment]
How in practice this is handled by password managers? I think most use browser extension (for web based logins) instead of keystroke injection, which should like more reliable option. Anyway, regardless of "typing" mechanism, this looks like a good idea. In fact I was considering something similar for some time already. My thoughts on it:
1. User gets some login prompt (application, web page)
2. User initiate password fill operation within that VM (either key combo handled within that VM, or a browser extension button)
3. That VM initiate RPC call, including request origin (application name, webapp URL) in call argument
4. Policy define whether given VM can access given credentials (based on call argument), you can also define which VM use which password-storing VM, if you want to separate them. This policy can also have action "ask", in which case you get a confirmation prompt, including target service name.
5. RPC call reaches password-storing VM and ask password manager for relevant credentials in some format. If there are multiple of them, another prompt (from password manager) appears.
6. Credentials are forwarded back to the original VM and injected into application it was asked for.

As said above - especially for web applications, browser extension should be better idea. Maybe it's even possible to use some existing one, and on a VM side pretending to be actual password manager but only forwarding requests to password-storing VM. This way we also get nice phishing prevention - since the URL will be transferred directly from the browser, all kind of attempts to fool the user about the actual URL will fail (like gmail-com.com, or using looking-alike unicode characters).

There could be also registration workflow that automatically set qrexec policy - on a basis that VM generating given credential can access it later. This can be a named DisposableVM. We do similar thing for our U2F integration.

Some of the above is inspired by https://github.com/QubesOS/qubes-app-u2f/

**Comment by Szewcson** [comment]
Don't know if it deserv separate issue but I want to propose adding simmilar mechanism for secret sevice API. This will allow user to enter only one password in "secrets" VM and run multiple apps like chromium, Spotify or Skype that want to access their secret at startup in separate VMs. I know that I can disable usage of password db encryption for many of that apps while not using builtin password managers but from UX/security point of view adding such mechanism seems to be a better option.

**Comment by fullTalgoRythm** [comment]
Although it would be nice if a browser extension is available. I think it might end up being a lot of work for someone to maintain. It also has several problems. A big reason I wanted this is for those obnoxious sites where you get pick ? chars from you pw. These sites are getting worse not better (mutlipe secrets and multiple prompts) and they all seem to have a unique UI which is actually designed to prevent you entering your password with a password manager. IMHO they want to shift risk for problems to the user so they can be blamed for using an insecure pw or not storing it in an encrypted format etc. This is however not a good reason not to use the appropriate software and fully randomize your pw as I think you are likely playing into their hands if you don't. We had systems like pub key authentication for a long time but they are never going to support that unless there was a massive online backlash against the way they do it now. I think the danger is that big tech is doing their FIDO alliance thing and I can't help thinking that these things normally mean we have to send personal details to those companies everytime and possibly get forced into using more of there products and services. I personally want to use less of those :)

Speaking personally again I would want support in all non browser places. These are also some of the worst offenders because sometimes the relevant software disables the clipboard. This is all done deliberatly in the name of 'security' and any argument to the contrary is a waste of time as it is mostly done again to shift blame / responsibility out of some organisation on to the individual etc. Hopefully the politics at play here are obvious but I feel any solution would need to work in these scenarios more than the others because they are deliberatly designed to waste time and weaken security so someone else can say 'it wasn't our fault our security is excelent'.

To me the most generic solution would therefore be best as it should support password entry in all software especially places where it can be a real nuisance to use a sensisble secure password. I think that is still essentially going to be keystroke injection but it would be nice if this was perhaps improved with an rpc policy and an API or similar so we can inject keystrokes into a target VM regardless of the input focus which can be changed by clicking out of a window etc. As per comments above it would be good to segment via the RPC policy which password store can give credentials to which VM. All the rest can be done with boiler plate although it would be nice if it was a defined solution I guess it depends on whether someone will be willing to develop it and it will properly work in all scenarios where it is required. Having a proper keystroke injection method would be better than trying to use qrexec-client (to run something like xdotool in the vm) as that would probably show in the process list with keystroke parameter visible on the client VM which is not ideal. If you run xdotool in dom0 it obviously is pretty much the same as using the keyboard afaik it just becomes forwarded X events in this instance but I guess that is more adhoc perhaps not as nice as a proper way to target a specific VM. If any API or similar had the ability to prevent spurious keyboard and mouse events during injection this would also be a benefit.

I have also noted that XFCE keyboard shortcut utility  / daemon appears to respond to keydown events rather than keyup event. So any adhoc scripts run in dom0 can have the problem that the keyup event for any modifier keys could corrupt the keystokes being injected. Having an extension of the qubes utils to allow adhoc scripts might be nice so we can provide our own solutions to these problems. This could give a bit more control over running an appropriate script in response to a hot key combination or whatever.

There is also the issue with keyboard injection that the keyboard layout etc. can cause symbol characters to be entered incorrectly. xdotool does seem able to workaround this but it can be flakey and it is not clear what the interaction between settings in the client VM and dom0 might do i.e. changing settings on one side but not the other may further confuse things.
